### ISSUE TITLE
[ENH] Hierarchical agglomerative estimation: E-Agglo

### ISSUE THREAD

**Issue by KatieBuc** [sktime/sktime#3397]
### Details 
E-Agglo is a nonparametric method for performing hierarchical agglomerative estimation of multiple change points based on [1][2]. Neighboring segments are then sequentially merged to maximize a goodness-of-fit statistic. The goodness-of-fit statistic is the "between-within distance" [3] among adjacent segments i.e.

![image](https://user-images.githubusercontent.com/34525024/189269160-707c101a-6344-456b-8571-048344348322.png)

where || . || is the Euclidean norm. Which is the scaled sample measure of divergence below

![image](https://user-images.githubusercontent.com/34525024/189269751-d45fa4e1-5278-48ba-b9f5-033cc9137051.png)

i.e.
``e(A, B) = mn/(m+n) * [minimum energy clustering]`` as evidenced in [4].

Note that minimum energy clustering can be computed with `scipy.stats.energy_distance(...)` [5].

#### References
[1] Matteson, David S., and Nicholas A. James. "A nonparametric approach for multiple change point analysis of multivariate data." Journal of the American Statistical Association 109.505 (2014): 334-345. URL: https://www.tandfonline.com/doi/full/10.1080/01621459.2013.849605
[2] James, Nicholas A., and David S. Matteson. "ecp: An R package for nonparametric multiple change point analysis of multivariate data." arXiv preprint arXiv:1309.3295 (2013). URL: https://arxiv.org/pdf/1309.3295.pdf
[3] Szekely, Gabor J., and Maria L. Rizzo. "Hierarchical clustering via joint between-within distances: Extending Ward's minimum variance method." Journal of classification 22.2 (2005): 151-184. URL: https://home.ipipan.waw.pl/j.mielniczuk/Szekely-Rizzo_2005.pdf
[4] https://en.wikipedia.org/wiki/Hierarchical_clustering#cite_ref-7
[5] https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.energy_distance.html

## Option 1: Interface sklearn.cluster.AgglomerativeClustering

[sklearn.cluster.AgglomerativeClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering) has parameters `linkage{‘ward’, ‘complete’, ‘average’, ‘single’}, default=’ward’`.

We need another linkage type not offered in this list. Therefore, this task involves writing another version of [ward_tree](https://github.com/scikit-learn/scikit-learn/blob/36958fb24/sklearn/cluster/_agglomerative.py#L169), with distnace between clusters `e(A,B)` from above instead of Ward's linkage defined as

![image](https://user-images.githubusercontent.com/34525024/189270752-245a4dc8-4747-4d28-959f-94f565bd0ba1.png)
(the visualisation of linkage parameters comes from [here](https://towardsdatascience.com/machine-learning-algorithms-part-12-hierarchical-agglomerative-clustering-example-in-python-1e18e0075019))

and interfacing `AgglomerativeClustering(ClusterMixin, BaseEstimator)` with this extended functionality.

### Dependencies

- [ ] `scipy.cluster.hierarchy.ward(X)` s called, where
![image](https://user-images.githubusercontent.com/34525024/189275455-a94e6ade-cd19-47a4-8948-7e31a596c10a.png)

which is (almost) the same as the desired incremental metric from [2]:
![image](https://user-images.githubusercontent.com/34525024/189278041-e2f4b226-fc38-4cd1-affa-37090de43bd9.png)

- [ ] `ward_dist` relies on c++ ` _hierarchical.compute_ward_dist` - can this be fully replaced with the scipy function? [5]
- [ ] `ward_dist` relies on c++  `_hierarchical._get_parents` - can this be called?

## Option 2: rewrite R code as stand alone

The R code is: https://github.com/cran/ecp/blob/master/R/e_agglomerative.R

- [x] getWithin, getBetween are C functions that need to be re-written
https://github.com/cran/ecp/blob/master/src/energyChangePoint.cpp

but maybe they can be replaced by `scipy.stats.energy_distance(...)` [5] - wouldn't take too long to verify.

**Comment by KatieBuc** [comment]
This is my hot take: this is confusing. I think the safest option would be to directly re-write the R implementation, write unit tests, then (potentially) refactor this logic to interface AgglomerativeClustering.

**Comment by KatieBuc** [comment]
Estimated time: 1-2 weeks until draft PR

**Comment by lmmentel** [comment]
@KatieBuc this is great, thanks. Would you at this point be able to draft e-Agglo in pseudocode so we could see what component we have and what would have to be implemented or interfaced?

**Comment by KatieBuc** [comment]
Pseudo code for the R-rewrite:

```
def process.data():

        #set up left and right neighbors, cluster size, checks
        C: getWithin() #create array of within distances 
        C: getBetween() #make distance matrix 
        #set initial GOF value
        #create matrix for change point progression
        #vector to specify the starting point of a cluster
```

```
def gof.update()
        #Function to calculate the new GOF value. The i argument is assumed to be the left 
        #cluster.

        #get new left and right clusters
        #get cluster sizes
        #add distance to new left cluster
        #add distance to new right cluster
```

```
def find.closest()
        #Determine which clusters will be merged. Returns a vector of length 3. The first 
        #element of the vector is the left cluster. The third element of the vector is the 
        #newest GOF value.
        
        #iterate to see how the GOF value changes
        for i in 1:K:
           gof.update() #get updated gof value
           #if better value found, update

```

```
def updateDistance():
        #Function to update the distance from the new cluster to the other clusters. 
        #i is assumed to be the left cluster. Also updates any other necessary 
        #information in the list (ret)
        
        #update left and right neighbors in ret
        #assign size to newly created cluster
        #update set of change points
        #update distances as per Lemma 3 in above comment
        for(k in 1:K):
              n3 = ret$sizes[k]
              n = n1+n2+n3
              ret$D[K+1,k] = ret$D[k,K+1] = ((n-n2)*ret$D[i,k]+(n-n1)*ret$D[j,k]-n3*ret$D[i,j])/n
```


```
def e.agglo():

    process.data() #initilize components
    # find which clusters optimize the GOF and then update the distances
    for K in N:...
        find.closest() #find which clusters to merge
        updateDistance() #update information after merger
    #penalize the GOF statistic
    #get the set of change points for the "best" clustering
    #create final membership vector
```

**Comment by KatieBuc** [comment]
If this were to be interfaced with [sklearn.cluster.AgglomerativeClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering), then all this would be refactored into an appropriated version of [ward_tree](https://github.com/scikit-learn/scikit-learn/blob/36958fb24/sklearn/cluster/_agglomerative.py#L169) within an instance of `AgglomerativeClustering(ClusterMixin, BaseEstimator):`. However, I'm still unsure of what would be useful in this class, since the fit approach is different i.e. 

```
def ward_tree(X, *, connectivity=None, n_clusters=None, return_distance=False):
    """Ward clustering based on a Feature matrix.
    Recursively merges the pair of clusters that minimally increases
    within-cluster variance.
    The inertia matrix uses a Heapq-based representation.
    This is the structured version, that takes into account some topological
    structure between samples.
```
how this `inertia` matrix relates to the greedy agglomerative procedure outlined here:

![image](https://user-images.githubusercontent.com/34525024/189785485-6aad98dc-5d34-4704-a8e1-764b986c48c9.png)

How to write the R in python is clear to me, how to interface this with a seemingly similar class is very unclear still.

**Comment by KatieBuc** [comment]
@lmmentel @Lovkush-A @fkiraly fyi

**Comment by KatieBuc** [comment]
In terms of re-writing this R code, a simplification of the function (which calls C functions) is:
```
install.packages('ecp')
library(ecp)

set.seed(1234)
X <- c(rnorm(15, mean = -6), 0, rnorm(16, mean = 6))

#Initialize all the necessary components of the list.
process.data = function(member,X,alpha){
  ret = NULL #the list with the necessary information
  u = unique(member)
  N = ret$N = length(u)#number of clusters

  for(i in 1:N)#relabel clusters
    member[member==u[i]] = i
  
  ret$sizes = numeric(2*N)
  
  for(i in 1:N)#calculate initial cluster sizes
    ret$sizes[i] = sum(member==i)
  
  #array of within distances
  within = numeric(N)
  for(i in 1:N)
    within[i] = ecp::getWithin(alpha,matrix(X[member==i,],ret$sizes[i],ncol(X)))
  
  ret$within = within
  
  #make distance matrix
  ret$D = matrix(Inf,nrow=2*N,ncol=2*N)
  for(i in 1:N)
    for(j in i:N)
      if(j != i)
        ret$D[i,j] = ret$D[j,i] = ecp::getBetween(alpha,matrix(X[member==i,],ret$sizes[i],ncol(X)),matrix(X[member==j,],ret$sizes[j],ncol(X))) - within[i] - within[j]
  
  diag(ret$D) = 0

  return(ret)
}
ret = process.data(1:32,as.matrix(X),1)
ret$D

```

which gives distance matrix:
![image](https://user-images.githubusercontent.com/34525024/189818648-6a6a0646-d0c3-4a55-b340-d0f252ef08d9.png)

The equivalent python code is:
```
def get_within(X, alpha):
    n = X.shape[0]
    return list(sum(np.power(np.sqrt(sum((X.iloc[i,]-X.iloc[j,])*(X.iloc[i,]-X.iloc[j,]))),alpha) for j in range(n))/(n**2) for i in range(n))

within = [get_within(X.loc[[i,]], alpha)[0] for i in range(N)]

def get_between(X, Y, alpha):
    n = X.shape[0]
    m = X.shape[0]
    return list(sum(np.power(np.sqrt(sum((X.iloc[i,]-Y.iloc[j,])*(X.iloc[i,]-Y.iloc[j,]))),alpha) for j in range(m))/(m * n) for i in range(n))

D = pd.DataFrame(index = range(N), columns = range(N))

for i in range(N):
    for j in range(N):
        between = get_between(X.loc[[i,]], X.loc[[j,]], alpha)[0]
        D.iloc[i,j] = D.iloc[j, i] = 2*between - within[i] - within[j]

```

which gives the expected distance matrix:
![image](https://user-images.githubusercontent.com/34525024/189818972-22d4ba56-d97f-4c73-8f8b-d02d00153494.png)

Also, note that this distance matrix could also be obtained using `scipy.stats.energy_distance()`
![image](https://user-images.githubusercontent.com/34525024/189819472-fa898d1f-35e6-4e7d-b795-eb9ff47a44fd.png)

which I think is the way to go with this.
